### PR TITLE
MinMaxCapability Wizardlets  Support

### PR DESCRIPTION
**Pull request by tharmaman**
- Modified the grant path method to include the  prereqs of minMaxChoices:

Format of a min max for getGrantPath
 [[prereqsChoiceA, choiceA], [prereqsChoiceB,choiceB], minMaxCapa]

Format of a min max for getGrantPath as a prereq for another  capability
[[[prereqsChoiceA, choiceA], [prereqsChoiceB,choiceB], minMaxCap],cap]

- The idea is that if there  is an array, if the last element is a MinMaxCapability then we can say this is an OR array
- Otherwise this is an AND array (i.e. for a  normal capability with just prereqs)

We designed this so it can work recursively both in the grantPath fetching and wizardlet generation

For AND  cases, prereq wizards will follow the isAvailable of their parent wizard (i.e. its easy to think of  this as the default case for a 1d grant path structure like [prereqAA, prereqAB, A] )

For OR cases, prereqs will follow what ever is chosen in the MinMaxCapabilityWizard since it uses the MultiChoiceView

Can support complex structures  if a prerequisite ends up being a minmax, then one of those min max choices ends up having prereqs or a minmax of its own.

Also included some other fixes as well that aren't off the top of my mind.

### CODE REVIEW

File: src/foam/nanos/crunch/MinMaxCapability.js [hunk 1]
```diff
@@ -74,6 +74,10 @@ foam.CLASS({
           }
         }
 
+        // TODO: Need to test if this works
+        System.out.println("numberGranted");
```
Comment (jlhughes) [inline src/foam/nanos/crunch/MinMaxCapability.js]: Can we make these logger.debug for now?

File: src/foam/nanos/crunch/ui/CapabilityWizardlet.js [hunk 2]
```diff
@@ -42,8 +42,8 @@ foam.CLASS({
         if ( ! this.of ) return null;
 
         var ret = this.of.getAxiomByName('capability') ?
-          this.of.create({ capability: this.capability }, this) :
-          this.of.create({}, this);
+          this.of.create({ capability: this.capability }, this.__subContext__) :
```
Comment (jlhughes) [inline src/foam/nanos/crunch/ui/CapabilityWizardlet.js]: Please bring this up and resolve with @kgrgreer

File: src/foam/nanos/crunch/MinMaxCapability.js (hunk 1 above)
Comment (tharmaman) [reply src/foam/nanos/crunch/MinMaxCapability.js]: Just removed them they were for testing

File: src/foam/u2/crunch/wizardflow/CreateWizardletsAgent.js [hunk 3]
```diff
@@ -38,55 +38,70 @@ foam.CLASS({
...
+      
+      array.slice(0, array.length - 1).forEach(
+        prereqCap => {
+          if ( Array.isArray(prereqCap) ){
+            updateUcjPromiseList = updateUcjPromiseList.concat(this.parseArrayToWizards(prereqCap, currentWizardlet));
+          } else {
+            var prereqWizardlet = prereqCap.wizardlet.cls_.create(
+              { 
+                capability: prereqCap, 
+                ...prereqCap.wizardlet.instance_ ,
```
Comment (kgrgreer) [inline src/foam/u2/crunch/wizardflow/CreateWizardletsAgent.js]: instance_ is meant to be an internal implementation detail, subject to change. Could you just not clone prereqCap.wizardlet the n set the capability? Wouldn't this accomplish the same thing more simply without exposing internals?

File: src/foam/u2/crunch/wizardflow/CreateWizardletsAgent.js [hunk 4]
```diff
@@ -38,55 +38,70 @@ foam.CLASS({
...
 
-                associatedEntity = capability.associatedEntity === foam.nanos.crunch.AssociatedEntity.USER ? this.subject.user : this.subject.realUser;
+      var currentCap = array[array.length - 1];
+      var currentWizardlet  = isOr 
+        ? foam.nanos.crunch.ui.MinMaxCapabilityWizardlet.create({
+            capability: currentCap,
+            ...currentCap.wizardlet.instance_,
+            choiceWizardlets: []
+          }, this.__subContext__)
+        : currentCap.wizardlet.cls_.create(
```
Comment (kgrgreer) [inline src/foam/u2/crunch/wizardflow/CreateWizardletsAgent.js]: same as below

File: src/foam/u2/crunch/wizardflow/CreateWizardletsAgent.js [hunk 5]
```diff
@@ -38,55 +38,70 @@ foam.CLASS({
...
-                  ...capability.wizardlet.instance_ 
-                }, this)
+    function parseArrayToWizards(array, parentWizardlet){
+      var isOr = foam.nanos.crunch.MinMaxCapability.isInstance(array[array.length - 1]) ? true : false;
+      var updateUcjPromiseList = [];
 
-                associatedEntity = capability.associatedEntity === foam.nanos.crunch.AssociatedEntity.USER ? this.subject.user : this.subject.realUser;
+      var currentCap = array[array.length - 1];
+      var currentWizardlet  = isOr 
+        ? foam.nanos.crunch.ui.MinMaxCapabilityWizardlet.create({
```
Comment (kgrgreer) [inline src/foam/u2/crunch/wizardflow/CreateWizardletsAgent.js]: same as below

File: src/foam/nanos/crunch/ui/CapabilityWizardlet.js (hunk 2 above)
Comment (kgrgreer) [reply src/foam/nanos/crunch/ui/CapabilityWizardlet.js]: I think this should just be 'this'.

File: src/foam/nanos/crunch/ServerCrunchService.java [hunk 6]
```diff
@@ -70,7 +70,38 @@ public List getCapabilityPath(X x, String rootId, boolean filterGrantedUCJ) {
...
+
+        // Manually grab the direct  prereqs to the  MinMaxCapability
+        List prereqs = ( (ArraySink) prerequisiteDAO
+          .where(AND(
+            EQ(CapabilityCapabilityJunction.SOURCE_ID, sourceCapabilityId),
+            NOT(IN(CapabilityCapabilityJunction.TARGET_ID, alreadyListed))
+          ))
+          .select(new ArraySink()) ).getArray();
+
+        for ( int i = prereqs.size() - 1; i >= 0; i-- ) {
```
Comment (kgrgreer) [inline src/foam/nanos/crunch/ServerCrunchService.java]: spaces fore ;'s

File: src/foam/nanos/crunch/ui/CapabilityWizardlet.js (hunk 2 above)
Comment (tharmaman) [reply src/foam/nanos/crunch/ui/CapabilityWizardlet.js]: Yeah so before @KernelDeimos and I were having difficulties with it  just being this,  but just tested it  right now with merge to foam master and it is no longer a problem.  Reverting back to just this.

File: src/foam/nanos/crunch/ServerCrunchService.java (hunk 6 above)
Comment (tharmaman) [reply src/foam/nanos/crunch/ServerCrunchService.java]: Done.

File: src/foam/u2/crunch/wizardflow/CreateWizardletsAgent.js (hunk 5 above)
Comment (tharmaman) [reply src/foam/u2/crunch/wizardflow/CreateWizardletsAgent.js]: Done. Don't even  need the ternary operator anymore if  I override 'wizardlet' property on MinMaxCapability to create  a MinMaxCapabilityWizardlet in the factory.

File: src/foam/u2/crunch/wizardflow/CreateWizardletsAgent.js (hunk 4 above)
Comment (tharmaman) [reply src/foam/u2/crunch/wizardflow/CreateWizardletsAgent.js]: Done.

File: src/foam/u2/crunch/wizardflow/CreateWizardletsAgent.js (hunk 3 above)
Comment (tharmaman) [reply src/foam/u2/crunch/wizardflow/CreateWizardletsAgent.js]: Done.

File: src/foam/u2/crunch/wizardflow/CreateWizardletsAgent.js [hunk 7]
```diff
@@ -38,55 +38,66 @@ foam.CLASS({
...
-            var choiceWizardlets = cap.slice(0, cap.length - 1).map(
-              capability => {
-                wizardlet = capability.wizardlet.cls_.create(
-                { 
-                  capability: capability, 
-                  isAvailable: false,
-                  ...capability.wizardlet.instance_ 
-                }, this)
+      var currentCap = array[array.length - 1];
+      var currentWizardlet  = currentCap.wizardlet.clone().clone().copyFrom(
```
Comment (kgrgreer) [inline src/foam/u2/crunch/wizardflow/CreateWizardletsAgent.js]: Why clone() twice?

Comment (tharmaman) [reply src/foam/u2/crunch/wizardflow/CreateWizardletsAgent.js]: oops just fixed

**Pull request merged**